### PR TITLE
Rewrite ILP algorithm employed in coffee mode

### LINKED ISSUE #135: Some gusto tests now take forever to optimise kernels
The ILP solver does the naive `2**N` thing.  We need to cut off somewhere.  For example, one kernel has 24 variables, so we're checking `2**24` options.

### PR DESCRIPTION
**Pull request by miklos1**
New version is believed to be much faster, and sensibly interruptible, however, this has not been thoroughly tested yet.

Fixes #135, closes #136.

### CODE REVIEW

File: tsfc/coffee_mode.py [hunk 1]
```diff
@@ -93,69 +84,63 @@ def monomial_sum_to_expression(monomial_sum):
...
-    return tuple(atomic for atomic in atomics if atomic in optimal_atomics)
+    optimal_solution = set(atomics)  # pessimal but feasible solution
+    solution = set()
+
+    max_it = 1 << 12
+    it = iter(range(max_it))
+
+    def solve(idx):
+        while idx < len(monomials) and solution.intersection(monomials[idx].atomics):
+            idx += 1
```
Comment (wence-) [inline tsfc/coffee_mode.py]: OK, find the first monomial that doesn't intersect with the current solution.

File: tsfc/coffee_mode.py [hunk 2]
```diff
@@ -93,69 +84,63 @@ def monomial_sum_to_expression(monomial_sum):
...
+
+        if idx < len(monomials):
+            if len(solution) < len(optimal_solution):
```
Comment (wence-) [inline tsfc/coffee_mode.py]: If we haven't intersected with all monomials less and the current solution is shorter than the optimal, then let's try.

File: tsfc/coffee_mode.py [hunk 3]
```diff
@@ -93,69 +84,63 @@ def monomial_sum_to_expression(monomial_sum):
...
+                for atomic in monomials[idx].atomics:
+                    solution.add(atomic)
+                    solve(idx + 1)
+                    solution.remove(atomic)
```
Comment (wence-) [inline tsfc/coffee_mode.py]: Try each atomic in the current monomial in turn.

File: tsfc/coffee_mode.py [hunk 4]
```diff
@@ -93,69 +84,63 @@ def monomial_sum_to_expression(monomial_sum):
...
+        else:
+            if cost(solution) < cost(optimal_solution):
+                optimal_solution.clear()
+                optimal_solution.update(solution)
```
Comment (wence-) [inline tsfc/coffee_mode.py]: We reached the end of the list of monomials so we have a feasible solution.  If it is cheaper than the current one, then accept.

File: tsfc/coffee_mode.py [hunk 5]
```diff
@@ -93,69 +84,63 @@ def monomial_sum_to_expression(monomial_sum):
...
+            next(it)
```
Comment (wence-) [inline tsfc/coffee_mode.py]: Keep track of the number of attempts.

**Comment by miklos1** [comment]
@wence-: do you want to test it with the Gusto test suite?

**Comment by wence-** [comment]
OK.

**Comment by wence-** [comment]
Seems good.

**Pull request merged**